ext/php5/serializer.c

```diff
@@ -445,14 +560,51 @@ void ddtrace_error_cb(DDTRACE_ERROR_CB_PARAMETERS) {
 
     bool is_fatal_error = type & (E_ERROR | E_CORE_ERROR | E_COMPILE_ERROR | E_USER_ERROR);
     if (EXPECTED(EG(active)) && EG(error_handling) == EH_NORMAL && UNEXPECTED(is_fatal_error)) {
-        ddtrace_exception_t *error = ddtrace_make_exception_from_error(DDTRACE_ERROR_CB_PARAM_PASSTHRU TSRMLS_CC);
-        ddtrace_span_fci *span = DDTRACE_G(open_spans_top);
-        while (span) {
-            ddtrace_span_attach_exception(span, error);
-            span = span->next;
+        /* If there is a fatal error in shutdown then this might not be an array
+         * because we set it to IS_NULL in RSHUTDOWN. We probably want a more
+         * robust way of detecting this, but I'm not sure how yet.
+         */
+        if (Z_TYPE(DDTRACE_G(additional_trace_meta)) == IS_ARRAY) {
+            dd_error_info error = dd_fatal_error(type, format, args);
+            dd_fatal_error_to_meta(&DDTRACE_G(additional_trace_meta), error);
+
+            ddtrace_span_fci *span;
+            for (span = DDTRACE_G(open_spans_top); span; span = span->next) {
+                if (span->exception || !span->span.span_data) {
+                    continue;
+                }
+                zval *meta = _read_span_property(span->span.span_data, ZEND_STRL("meta") TSRMLS_CC);
+                if (!meta) {
+                    continue;
+                }
+
+                if (Z_TYPE_P(meta) == IS_NULL) {
+                    // Unlike PHP 7, we cannot mutate the SpanData properties outside of the zend_*_property APIs
+                    zval *zmeta;
+                    MAKE_STD_ZVAL(zmeta);
+                    array_init_size(zmeta, ddtrace_num_error_tags);
+                    zend_update_property(ddtrace_ce_span_data, span->span.span_data, ZEND_STRL("meta"),
+                                         zmeta TSRMLS_CC);
+                    zval_ptr_dtor(&zmeta);
+
+                    meta = _read_span_property(span->span.span_data, ZEND_STRL("meta") TSRMLS_CC);
+                }
+                if (Z_TYPE_P(meta) == IS_ARRAY) {
+                    dd_fatal_error_to_meta(meta, error);
+                }
```

Comment:
Let me make sure I understand this a bit:

1. We update the property with an empty array with a capacity of `ddtrace_num_error_tags` or 3.
2. We then read that property back to get the actual zval of the empty array.
3. Now we append the fatal error info to the zval of the empty array.

Why don't we write the fatal error info before we update the property?

Reply:
> Why don't we write the fatal error info before we update the property?

If I understand your question correctly, we cannot do this since `SpanData::$meta` might be an existing array that was set from a prehook tracing closure. Is that what you are getting at?

Reply:
Yes, it might be an existing array, but the `Z_TYPE_P(meta) == IS_NULL` already rules that out for this branch. I guess I'd have to see the full code of splitting those cases out a bit more to see if I like that or what is written better, but minimizing calls into Zend APIs seems preferable.

Reply:
Ah! I see what you're saying. Yes, we can remove having to read the property again by adding the error info to the initialized array before updating the property. Good catch. I just updated this.